# Notebook 02 — Preprocessing

Pipeline: contraction normalization → noise removal → tokenise → remove html-junk tokens → POS tag → lemmatise → stopword removal → review fragments → store

Text-cleaning functions live in `src/preprocessing.py`, imported below. Each step still runs in its own cell so the reasoning stays visible.

In [1]:
import sys
sys.path.insert(0, '..')

import pickle
import pandas as pd
import nltk
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

from src.preprocessing import (
    normalize, clean, tokenize_raw, remove_html_junk,
    get_wordnet_pos, lemmatize, get_stop_words, remove_stopwords,
    NEGATION_WORDS, preprocess,
)

DATASET_PATH = "../data/raw/Emotion_classify_Data.csv"
df = pd.read_csv(DATASET_PATH)

# toggle these to test each step's effect
APPLY_LEMMATIZATION = True
APPLY_STOPWORDS = True
PROTECT_NEGATION = True

## 2.0 Drop contradictory duplicates

EDA (Notebook 01) identified 3 comment pairs that appear under two different labels — always `anger` vs `fear`. All 6 rows are dropped before the pipeline runs to prevent any of them leaking across the train/val/test split.

In [2]:
conflict_comments = (
    df.groupby("Comment")["Emotion"]
    .nunique()
    .loc[lambda s: s > 1]
    .index
)
df = df[~df["Comment"].isin(conflict_comments)].reset_index(drop=True)
print(f"Rows after dropping contradictory duplicates: {len(df)}")

Rows after dropping contradictory duplicates: 5931


**Result:** 5,931 rows (6 dropped — the 3 contradictory pairs).

## 2.1 Contraction normalization + noise removal

The raw CSV is already lowercased and stripped of punctuation — negation arrives pre-merged (`dont`, `cant`, `didnt`, ...) instead of with an apostrophe. `normalize()` expands both the merged and apostrophe'd forms to the same full words (`"do not"`) before the alpha-only strip, so negation always ends up as a literal `not` token.

A few merged forms (`ill`, `hell`, `well`, `its`, `were`, `id`) are left alone since collapsing them would collide with a real word.

`Comment` stays untouched — the normalized/cleaned text goes into separate columns.

In [3]:
df["Normalized_Comment"] = df["Comment"].apply(normalize)
df["Cleaned_Comment"] = df["Normalized_Comment"].apply(clean)
df["Cleaned_Comment"]


0       i seriously hate one subject to death but now ...
1                    i am so full of life i feel appalled
2       i sit here to write i start to dig out my feel...
3       i have been really angry with r and i feel lik...
4       i feel suspicious if there is no one outside l...
                              ...                        
5926                   i begun to feel distressed for you
5927    i left feeling annoyed and angry thinking that...
5928    i were to ever get married i d have everything...
5929    i feel reluctant in applying there because i w...
5930    i just wanted to apologize to you because i fe...
Name: Cleaned_Comment, Length: 5931, dtype: str

**Result:** `Comment` unchanged. `Normalized_Comment` holds the expanded text (`"i dont feel happy"` → `"i do not feel happy"`). `Cleaned_Comment` strips that to `[a-zA-Z\s]` only.

## 2.2 Tokenisation

Split `Cleaned_Comment` into tokens with `nltk.word_tokenize()`, stored in `tokens`. Contractions are already handled in 2.1, so this is just whitespace/punctuation splitting here.

In [4]:
df['tokens'] = df['Cleaned_Comment'].apply(tokenize_raw)
df['tokens'].head(5)

0    [i, seriously, hate, one, subject, to, death, ...
1       [i, am, so, full, of, life, i, feel, appalled]
2    [i, sit, here, to, write, i, start, to, dig, o...
3    [i, have, been, really, angry, with, r, and, i...
4    [i, feel, suspicious, if, there, is, no, one, ...
Name: tokens, dtype: object

## 2.3 Remove html-junk tokens

Fragments like `www`, `href`, `src` survive the alpha-only strip since they're purely alphabetic. Filtered out at the token level by an explicit list, not a broader regex.

In [5]:
df["Cleaned_Tokens"] = df["tokens"].apply(remove_html_junk)
df["Cleaned_Tokens"].head(5)

0    [i, seriously, hate, one, subject, to, death, ...
1       [i, am, so, full, of, life, i, feel, appalled]
2    [i, sit, here, to, write, i, start, to, dig, o...
3    [i, have, been, really, angry, with, r, and, i...
4    [i, feel, suspicious, if, there, is, no, one, ...
Name: Cleaned_Tokens, dtype: object

**Result:**

## 2.4 POS tagging

Tagging runs before stopword removal — the tagger uses auxiliary verbs and articles as context, so removing them first would hurt tagging (and lemmatisation) accuracy.

`nltk.pos_tag()` returns Penn Treebank tags (`VBG`, `NN`, `JJ`...), mapped to WordNet's simpler tags next.

In [6]:
df['POS_Tags'] = df['Cleaned_Tokens'].apply(lambda x: nltk.pos_tag(x))
df['POS_Tags'].head(5)

0    [(i, NN), (seriously, RB), (hate, VB), (one, C...
1    [(i, NN), (am, VBP), (so, RB), (full, JJ), (of...
2    [(i, JJ), (sit, NN), (here, RB), (to, TO), (wr...
3    [(i, NNS), (have, VBP), (been, VBN), (really, ...
4    [(i, NN), (feel, VBP), (suspicious, JJ), (if, ...
Name: POS_Tags, dtype: object

In [7]:
print(f'Number of unique tokens: {len(df["POS_Tags"].explode().unique())}')

Number of unique tokens: 11409


**Result:**

## 2.5 Lemmatisation

`get_wordnet_pos()` maps Penn tags to WordNet's tag set so lemmatisation is tag-aware (`"feeling"` as a verb → `"feel"`, as a noun stays `"feeling"`).

Known gap: WordNet has no entry for irregular verb forms, so `"felt"`, `"was"`, `"went"` survive unchanged even when verb-tagged (checked against spaCy below).

Controlled by `APPLY_LEMMATIZATION`.

In [8]:
# get_wordnet_pos is imported from src.preprocessing (see the imports cell) rather than
# redefined here. Quick sanity check that the mapping behaves as described above:
for tag in ['JJ', 'VBG', 'VBD', 'NN', 'RB', 'IN']:
    print(f"{tag} -> {get_wordnet_pos(tag)}")

JJ -> a
VBG -> v
VBD -> v
NN -> n
RB -> r
IN -> n


**Result:** mapping checks out. `.apply()` runs in 2.5b.

### 2.5a Before/after verification

Quick sanity check before running on all 5,931 rows, plus a spaCy comparison to confirm the irregular-verb gap is WordNet's, not a bug.

In [9]:
import spacy
nlp = spacy.load("en_core_web_sm")
doc = nlp("i went to school and had an argument with my friend. I was very angry and frustrated.")
for token in doc:
    print(token.text, token.lemma_)

i I
went go
to to
school school
and and
had have
an an
argument argument
with with
my my
friend friend
. .
I I
was be
very very
angry angry
and and
frustrated frustrate
. .


**Result:** confirmed — `felt`/`went`/`had`/`was` are a WordNet gap, not a logic bug.

### 2.5b Run on full dataset

Apply lemmatisation to all rows, then check vocab reduction and spot-check `feeling`/`feels`/`felt`.

In [10]:
if APPLY_LEMMATIZATION:
    df['Lemmatized_Tokens'] = df['Cleaned_Tokens'].apply(lemmatize)
else:
    df['Lemmatized_Tokens'] = df['Cleaned_Tokens']

before = len(df['Cleaned_Tokens'].explode().unique())
after = len(df['Lemmatized_Tokens'].explode().unique())
print(f"Unique tokens before lemmatization: {before:,}")
print(f"Unique tokens after lemmatization:  {after:,}")
print(f"Vocabulary reduction: {(before - after) / before:.0%}")

Unique tokens before lemmatization: 8,931
Unique tokens after lemmatization:  7,273
Vocabulary reduction: 19%


In [11]:
target_words = ["feeling", "feels", "felt"]

mask = df['Cleaned_Tokens'].apply(lambda tokens: any(target_word in tokens for target_word in target_words))

matches = df[mask]
print(f"Rows containing '{target_words}': {len(matches)}")

for idx in matches.index[:20]:
    original_tokens = df.loc[idx, 'Cleaned_Tokens']       
    lemmatized_tokens = df.loc[idx, 'Lemmatized_Tokens']
    
    if "feeling" in original_tokens :
        position = original_tokens.index("feeling")
    elif "feels" in original_tokens :
        position = original_tokens.index("feels")
    elif "felt" in original_tokens :
        position = original_tokens.index("felt") 
    else :
        None
    
    print(f"Row {idx}:")
    print(f"  original word:   {original_tokens[position]}")
    print(f"  became:          {lemmatized_tokens[position]}")

Rows containing '['feeling', 'feels', 'felt']': 2068
Row 7:
  original word:   feeling
  became:          feel
Row 11:
  original word:   feeling
  became:          feel
Row 13:
  original word:   feeling
  became:          feel
Row 18:
  original word:   feeling
  became:          feel
Row 20:
  original word:   feeling
  became:          feeling
Row 22:
  original word:   feeling
  became:          feel
Row 24:
  original word:   feeling
  became:          feeling
Row 25:
  original word:   feeling
  became:          feeling
Row 27:
  original word:   feeling
  became:          feeling
Row 28:
  original word:   feeling
  became:          feel
Row 34:
  original word:   feeling
  became:          feel
Row 36:
  original word:   feeling
  became:          feel
Row 38:
  original word:   felt
  became:          felt
Row 39:
  original word:   felt
  became:          felt
Row 40:
  original word:   feeling
  became:          feel
Row 41:
  original word:   feeling
  became:          fee

In [12]:
before = len(df['POS_Tags'].explode().unique())
after = len(df['Lemmatized_Tokens'].explode().unique())
print(f"Unique tokens before lemmatization: {before}")
print(f"Unique tokens after lemmatization: {after}")

Unique tokens before lemmatization: 11409
Unique tokens after lemmatization: 7273


**Result:** vocabulary reduced 19% (8,931 → 7,273 tokens). `feeling` → `feel` when verb-tagged, stays `feeling` as a noun — both correct. `felt` stays `felt` — WordNet gap confirmed.

## 2.6 Stopword removal

Removed after lemmatisation, using a plain Python set — no vectoriser. Negation words (`not`, `no`, `nor`, `never`, ...) are excluded from the stopword set so they survive, which only works because 2.1 already normalised fused negation to `not`.

Controlled by `APPLY_STOPWORDS`, with `PROTECT_NEGATION` toggling the negation exception on its own.

In [13]:
stop_words = get_stop_words(protect_negation=PROTECT_NEGATION)

if APPLY_STOPWORDS:
    df["filtered_tokens"] = df['Lemmatized_Tokens'].apply(
        lambda tokens: remove_stopwords(tokens, protect_negation=PROTECT_NEGATION)
    )
else:
    df["filtered_tokens"] = df['Lemmatized_Tokens']

mask = df['Lemmatized_Tokens'].apply(lambda tokens: any(word in tokens for word in NEGATION_WORDS))

matches = df[mask]
print(f"Rows containing a negation word: {len(matches)}")

for idx in matches.index[:5]:
    original = df.loc[idx, 'Lemmatized_Tokens']
    filtered = df.loc[idx, 'filtered_tokens']

    print(f"Row {idx}:")
    print(f"  original:  {original}")
    print(f"  filtered:  {filtered}")

Rows containing a negation word: 1399
Row 2:
  original:  ['i', 'sit', 'here', 'to', 'write', 'i', 'start', 'to', 'dig', 'out', 'my', 'feeling', 'and', 'i', 'think', 'that', 'i', 'be', 'afraid', 'to', 'accept', 'the', 'possibility', 'that', 'he', 'might', 'not', 'make', 'it']
  filtered:  ['sit', 'write', 'start', 'dig', 'feeling', 'think', 'afraid', 'accept', 'possibility', 'might', 'not', 'make']
Row 4:
  original:  ['i', 'feel', 'suspicious', 'if', 'there', 'be', 'no', 'one', 'outside', 'like', 'the', 'rapture', 'have', 'happen', 'or', 'something']
  filtered:  ['feel', 'suspicious', 'no', 'one', 'outside', 'like', 'rapture', 'happen', 'something']
Row 10:
  original:  ['i', 'feel', 'a', 'bit', 'like', 'franz', 'liebkind', 'in', 'the', 'producer', 'not', 'many', 'people', 'know', 'it', 'but', 'the', 'fuhrer', 'be', 'a', 'terrific', 'dancer']
  filtered:  ['feel', 'bit', 'like', 'franz', 'liebkind', 'producer', 'not', 'many', 'people', 'know', 'fuhrer', 'terrific', 'dancer']
Row 11:


**Result:** 1,399 rows keep a negation token (up from 932 before the 2.1 fix). E.g. row 13, "i do not always find myself feel thankful" → keeps `not` — without it the sentence reads as positive.

### 2.6a Verification

Run the full pipeline on a few standalone sentences to confirm contractions and negation survive end-to-end.

In [14]:
demo_sentences = [
    "I don't feel happy",
    "I didn't enjoy this",
    "I'm terrified",
    "I can't stop worrying",
]

for sentence in demo_sentences:
    normalized = normalize(sentence)
    filtered = preprocess(sentence, APPLY_LEMMATIZATION, APPLY_STOPWORDS, PROTECT_NEGATION)
    print(f"{sentence!r}")
    print(f"  normalized: {normalized!r}")
    print(f"  filtered:   {filtered}")


"I don't feel happy"
  normalized: 'i do not feel happy'
  filtered:   ['not', 'feel', 'happy']
"I didn't enjoy this"
  normalized: 'i did not enjoy this'
  filtered:   ['not', 'enjoy']
"I'm terrified"
  normalized: 'i am terrified'
  filtered:   ['terrify']
"I can't stop worrying"
  normalized: 'i can not stop worrying'
  filtered:   ['not', 'stop', 'worrying']


**Result:**
```
"I don't feel happy"    -> ['not', 'feel', 'happy']
"I didn't enjoy this"   -> ['not', 'enjoy']
"I'm terrified"         -> ['terrify']
"I can't stop worrying" -> ['not', 'stop', 'worrying']
```

Negation survives in all cases. `worrying` stays a noun here since the tagger has little context — same kind of tagger-dependent quirk as 2.5's irregular verbs.

## 2.7 Flag and review low-token-count rows

Flagging by token count only finds candidates — it's not itself a reason to drop a row ("terrified" alone would also hit `len <= 1` but is a valid example). Each flagged row gets checked against its raw text before anything is dropped.

In [15]:
mask = df['filtered_tokens'].apply(lambda tokens: len(tokens) <= 1)

flagged = df[mask]
print(f"Candidate rows with 0 or 1 tokens remaining: {len(flagged)}")
for idx in flagged.index:
    print(f"Row {idx}: filtered_tokens={df.loc[idx, 'filtered_tokens']}  |  raw: {df.loc[idx, 'Comment']!r}"
    )

Candidate rows with 0 or 1 tokens remaining: 6
Row 309: filtered_tokens=['sadness']  |  raw: 'as in sadness a'
Row 896: filtered_tokens=['army']  |  raw: 'in the army'
Row 1291: filtered_tokens=['exam']  |  raw: 'before an exam which i'
Row 3030: filtered_tokens=['sweden']  |  raw: 'in sweden'
Row 4740: filtered_tokens=['school']  |  raw: 'at school'
Row 5363: filtered_tokens=['lecture']  |  raw: 'during lectures'


### 2.7a Per-row review

| Row | Raw comment | Judgment |
|---|---|---|
| 309 | "as in sadness a" | Truncated mid-sentence. `sadness` survives only by accident of where it cut off. |
| 896 | "in the army" | Bare context phrase, no emotional content. |
| 1291 | "before an exam which i" | Cuts off mid-clause. |
| 3030 | "in sweden" | Bare location phrase. |
| 4740 | "at school" | Bare location phrase. |
| 5363 | "during lectures" | Bare temporal phrase. |

All 6 are truncated fragments — dropped for that reason, not for their token count.

In [16]:
# all 6 reviewed in 2.7a — real fragments, not valid short examples
fragments_to_drop = [309, 896, 1291, 3030, 4740, 5363]

df = df.drop(index=fragments_to_drop)
df = df.reset_index(drop=True)
print(f"Rows after dropping reviewed fragments: {len(df)}")

Rows after dropping reviewed fragments: 5925


**Result:** 6 rows dropped after review, all confirmed fragments. 5,925 rows remain.

## 2.8 Label encoding

Convert string labels to integers. Save the encoder to preserve the class mapping for decoding predictions at inference time.

> Print the mapping explicitly: `dict(enumerate(le.classes_))`

In [17]:
le = LabelEncoder()
df['Label'] = le.fit_transform(df['Emotion'])
print(dict(enumerate(le.classes_)))

{0: 'anger', 1: 'fear', 2: 'joy'}


**Result:** Labels encoded as `{0: 'anger', 1: 'fear', 2: 'joy'}`.

## 2.9 Train / validation / test split

Stratified 70 / 15 / 15 split to preserve class proportions in all three sets.

In [18]:
X_train, X_temp, y_train, y_temp = train_test_split(df[['Comment', 'filtered_tokens']], df['Label'], test_size=0.3, train_size=0.7, random_state=42, shuffle=True, stratify=df['Label'])
X_valid, X_test, y_valid, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42, shuffle=True, stratify=y_temp)

**Result:** 5,925 rows split 70/15/15 into train (4,147), validation (889), and test (889) sets.

## 2.10 Verify split balance

Confirm class ratios held across all three splits — a failed stratification silently skews everything downstream.

In [19]:
print(f"train set proportion: {y_train.value_counts(normalize=True) * 100}")
print(f"validation set proportion: {y_valid.value_counts(normalize=True) * 100}")
print(f"test set proportion: {y_test.value_counts(normalize=True) * 100}")

train set proportion: Label
2    33.735230
0    33.662889
1    32.601881
Name: proportion, dtype: float64
validation set proportion: Label
2    33.745782
0    33.633296
1    32.620922
Name: proportion, dtype: float64
test set proportion: Label
2    33.745782
0    33.633296
1    32.620922
Name: proportion, dtype: float64


**Result:** Class proportions hold across all three splits — joy ~33.7%, anger ~33.6%, fear ~32.6% in train, validation, and test alike. Stratification worked as intended.

Validation and test proportions match to 6 decimal places since both came from an even 50/50 split of the same 889-row temp set — not a bug, just two equal-sized stratified halves of the same data.

## 2.11 Save

Persist the split dataframes and label encoder so Notebook 03 can load clean data without rerunning this pipeline.

`assign()` recombines `Comment`/`filtered_tokens` with `Label` per split, since `train_test_split` keeps X/y separate. Saved as separate `.pkl` files per split so each stage loads only what it needs.

In [20]:
train_df = X_train.assign(Label=y_train)
valid_df = X_valid.assign(Label=y_valid)
test_df = X_test.assign(Label=y_test)

In [21]:
to_save = {
    'train.pkl': train_df,
    'valid.pkl': valid_df,
    'test.pkl': test_df,
    'label_encoder.pkl': le,
}
for filename, obj in to_save.items():
    with open(f'../data/processed/{filename}', 'wb') as f:
        pickle.dump(obj, f)

**Result:** Saved `train.pkl` (4,147 rows), `valid.pkl` (889 rows), `test.pkl` (889 rows), and `label_encoder.pkl` to `../data/processed/`. Each split table carries `Comment`, `filtered_tokens`, and `Label` together.